clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -786,6 +809,29 @@ private void handleAbortedSends(List<ClientResponse> responses) {
         abortedSends.clear();
     }
 
+    /**
+     * Handle socket channel connection timeout. The timeout will hit iff a connection
+     * stays at the ConnectionState.CONNECTING state longer than the timeout value,
+     * as indicated by ClusterConnectionStates.NodeConnectionState.
+     *
+     * @param responses The list of responses to update
+     * @param now The current time
+     */
+    private void handleTimedOutConnections(List<ClientResponse> responses, long now) {
+        Set<String> connectingNodes = connectionStates.connectingNodes();
+        for (String nodeId : connectingNodes) {
+            if (connectionStates.isConnectionSetupTimeout(nodeId, now)) {
+                this.selector.close(nodeId);
+                log.debug(
+                    "Disconnecting from node {} due to socket connection setup timeout. " +
+                    "The timeout value is {} ms.",
+                    nodeId,
+                    connectionStates.connectionSetupTimeoutMs(nodeId));
+                processDisconnection(responses, nodeId, now, ChannelState.LOCAL_CLOSE);
```

Comment:
@guozhangwang Thanks for reporting the exception in this code.
@d8tltanc @dajac This code segment is unsafe, we are removing `node` from `connectingNodes` in p`processDisconnection()` while iterating over the set. We must be missing a test too (or we have a test with only one connection).